058-zk-kraft-migration.md

```diff
@@ -0,0 +1,437 @@
+The `strimzi.io/kraft: disabled` annotation is already applied on the `Kafka` custom resource.
+The FSM is in the `KRaftEnablingMigration` state (as per `Kafka.status.metadataState` field) and keep staying in this state until the user deletes the controllers pool(s) to rollback (or re-apply `strimzi.io/kraft: migration` to move forward with migration).
+
+**_Trigger_**
+
+The user deletes the `KafkaNodePool` custom resource(s) defining the controllers already running with ZooKeeper migration enabled.
+
+**_Actions_**
+
+In this case, the operator starts a reconciliation an deletes the pods running the controllers.
+The current reconciliation ends.
+
+**_Status changes_**
+
+The FSM moves back to the `ZooKeeper`, so actually at the beginning of the Phase 1.
+There is no more migration in progress and the Apache Kafka cluster is running in ZooKeeper mode.
+
+#### Phase 3 and 4: Migrating brokers and controllers to full KRaft
+
+At the beginning of this phase, the user has the last opportunity to rollback the migration process.
+
+**_Preconditions_**
+
+The FSM is in the `KRaftDualWriting` state (as per `Kafka.status.metadataState` field).
+
+**_Trigger_**
+
+The user can rollback the process by applying the `strimzi.io/kraft: disabled` annotation on the `Kafka` custom resource.
+The operator reports a warning condition in the `Kafka` resource status if the ZooKeeper metadata are behind the KRaft ones, which means metadata loss if the user decides to rollback (see `ZkWriteBehindLag` metric description).
+
+**_Actions_**
+
+In this case, the operator starts a reconciliation and, via the `KafkaReconciler` component, configures the brokers with no ZooKeeper migration flag, and no KRaft controllers quorum connection information.
+Finally, the operator rolls the brokers, via the `KafkaRoller`, and waits for them to be ready again.
+The current reconciliation ends.
+
+**_Status changes_**
+
+The FSM moves back to the `KRaftEnablingMigration`, so actually at the beginning of the Phase 2.
+
+---
+
+If the user doesn't rollback the migration, next step is to finalize the migration process. This step is irreversible.
+
+**_Preconditions_**
+
+The FSM is in the `KRaftDualWriting` state (as per `Kafka.status.metadataState` field) and the operator expects the user desire to finalize the migration, having a full KRaft based cluster and ZooKeeper out of the picture.
+
+**_Trigger_**
+
+The user applies the `strimzi.io/kraft: enabled` annotation to the `Kafka` custom resource.
+
+**_Actions_**
+
+The operator starts a reconciliation and, via the `KafkaReconciler` component, updates the brokers' configuration by removing the ZooKeeper migration flag and connection to the ZooKeeper ensemble.
+It also has to replace the `broker.id` with `node.id`.
+More specifically, the brokers are automatically updated with the following configuration (the user cannot set them manually).
+
+```shell
+# replace broker.id with node.id
+node.id=0
+
+# Remove the migration enabled flag
```

Comment:
Are the settings actually commented out or are they removed from the configuration?

Reply:
They are removed. The snippet has them commented out just to explain what's going to be "disabled". I followed the way the official Kafka documentation does, which is understandable imho.